sql: couldn't decode type anyelement

roachtest.sqlsmith/setup=rand-tables/setting=no-ddl [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4807868&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4807868&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-ddl) on release-22.1 @ [1ec23d5938e3e0a2dc4deff3fc05c2ea6b33cd80](https://github.com/cockroachdb/cockroach/commits/1ec23d5938e3e0a2dc4deff3fc05c2ea6b33cd80):


```
									'6301b3ec-382c-4dfd-a7c5-f1e4a02d59e1':::UUID,
									'0101000040789B257A4549F54169F82DD56D8AFFC11CF0A99E6EAAE041':::GEOMETRY,
									true
								),
								(
									'00000000-0000-0000-0000-000000000000':::UUID,
									'010100008066CA77E25849FBC1A04CB92A934CFBC10C70B2D3F859E741':::GEOMETRY,
									false
								)
						)
							AS tab_16433 (col_27119, col_27120, col_27121)
				),
			with_5407 (col_27122)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								('["foobar", [false, {"k6$K": 0.2494717786875567}, {}], {}, [{}], {}]':::JSONB),
								(e'{"2bhD\\\\pAJ4HO": [], "a": null, "foobar": []}':::JSONB),
								(
									'[{"baz": [";:Tog*a9(~``#", true, []], "c": [[2.322301667738751], {}], "c*6A!ev:": [0.1385869005111981, {"foobar": {}}, [], []]}, null]':::JSONB
								),
								('[null, [null, [], null], []]':::JSONB),
								('["baz", [0.009203375318097229], [{"c": {}}, []]]':::JSONB)
						)
							AS tab_16434 (col_27122)
					UNION ALL
						SELECT
							*
						FROM
							(
								VALUES
									(
										e'{"HU}JCKdTa< ": {";L/o``XA%LK<y": {"c": {"X%-+": {}}}, "Cf\\"Vp_KDf": {}, "b": [], "foo": [[[], {"c7y": null}]]}, "b": [{"WBp6++": {"JX|Ni1lP": null}}, []], "r[X[0": 0.32080903806026484}':::JSONB
									),
									(NULL)
							)
								AS tab_16435 (col_27123)
				)
		SELECT
			(ARRAY[(-14474):::INT8,28946:::INT8,(-1):::INT8,4766:::INT8,(-8036):::INT8,11874:::INT8], ARRAY[('1993-09-11':::DATE, '0107000060E6100000090000000106000060E610000009000000010300004001000000090000008AAE387139E662C07082F1F9B91239C0887C4D8F4A06FDC1CBB56417A80950C0E6EAC432B95751C070BA90751540E74108000E2CD1EF49C01A4C2F0D22CC54C0E80FA386DAE9E141B8AF52CABC333740A4FB0258860F50C0807C2EB9984A914158143BC15DBF5A40D6AA98752A9F55C05054AC5AA72CF541D000DD8D6ECA22C07814E3052B015640901916D66AB8DDC18ABBC9D62AB94BC04865366DB3634A4070663C34A146E841BC57AB77CA2958C0000BF3C8CC315440333DCA8464B5FDC18AAE387139E662C07082F1F9B91239C0887C4D8F4A06FDC1010300004001000000060000003E1CC66B15E352C08C41B31321B43840A889F5329403E641887CC28DADDD5AC0456196B14FBF46C002C82AA26DFEF74150489A53D9175B409AF6960C583551409819897AA4EBF0419041251DB9D05C4040A1AE6B529F54408CBA16C46B54F44112279CA989F759C0C8D46CAF20F44D401348975E4CA8F0C13E1CC66B15E352C08C41B31321B43840A889F5329403E6410103000040010000000C0000001137756AB2A059C040F648E4BD461440E4E926CF8F5AFFC133F6F591313255C06C85EF7B332432C01874D8FBA943F641448AEEB0B0FA62C0D107CAF8D6DA50C008C112533FB9E741E0A0E11D20454AC08EC8A742E4753BC02A5018A7B7E8F5C18821D76DF28E4E40B345CF42945147C0C0E4BFB2A0C3FD41DA6566085C856240C4A9189F62FC3BC06EE37DBA0A07004272A4DDFCEC8A6040F8EA7CB03E354740257248D7BE5500C268A62B0FFA3D5440AC648214DD5D4F4000518236A2BA9AC1C011D308366709C094BF9FDD2947424080228C533929AA41009E6C2E89342DC054B8320E1BC44440D44B3BF3F8EDFC4138E1FA25FFB457C0C6FC8186B89A5140E1162F68250F00C21137756AB2A059C040F648E4BD461440E4E926CF8F5AFFC101030000400100000007000000B0BB2D9072DA23C06A006FE9BF983CC03F1ED37658C8FDC12C9B6DF1F98A5F40E5DB5C8626AB52C05C8DD5FC8AFDFBC18079FF1784504240C8D71B87BA094F402CF350AFE154FF4150EFE43434D522C0441CE0C6627C5140A8439D3F9947F9419739C1B749C965C0964C35D999DC514000CF0B0C1D7A82410CAB6FB63F8D5BC0C0DEF9AACAA84640181AB6C8942CF0C1B0BB2D9072DA23C06A006FE9BF983CC03F1ED37658C8FDC101030000400100000009000000B78272C72BF760C010E96B68617127406EC33809C49901C280BC8BB61BEA03C0E0AB608020EF2DC04C1288A4E9AADBC1B8A16153289C5F4037D1DEEDB5A953C072F94E0AE9E9EBC1F850885A51A9594040340B14A6063840E8B95C3735A3F3C1F439F9386D0B5740EC2178C3139249407C77ED009564FB419428E510686F46C020010FECF3F44B40414D491FDBD901C2DC61A851E8035AC0262BFF70C98C5340E8A9AB4FDF22F4C184DA0AD2BDB444C08895C0ADF07F44404465800CEF1AEF41B78272C72BF760C010E96B68617127406EC33809C49901C20103000040010000000500000049648420BFB662C053F63E991A014DC00075A28D55EBBDC101859A9FC79C65C04CD9D534F73D51C050346A0ABBFBD6416447FF5110DE48C0F94A1C2BEC5853C048648FEA993CDAC1B08BF4629F3151406061F829CE9628C046984986161FF4C149648420BFB662C053F63E991A014DC00075A28D55EBBDC10103000040010000000800000014FB8E73F48A42C0E8B38F2F521A51C024F2C13A73D7EAC1A0ABA7890F6221C0CE2C937357D941C0BC7E520B1660F8417C879661CA8D494096615FB7EFE451C04D2824BF9CABF6C1F0B42270C1FC32C0FC53418899F74040DCE9ABA97637FC41509A41B8D5BA2BC07C339BB7AEDE5240401A16DB3AD4F6C18C10E69482F854C0A6D9557B4A68524040142453DCD5CDC13D0679F92EA956C0526F8A985A425440E8E4311D0D95DD4114FB8E73F48A42C0E8B38F2F521A51C024F2C13A73D7EAC10103000040010000000800000048127B22928A5BC03699161014B445C0FF7ABCF1E7CAF1C1640D1CA436A55EC096457E1E8C2A4FC0B073242C5980FF413013097508754540944C196209EC4DC098CE694A4158F44108490FE1C78964403CCA187D57FB4DC06E415F4476C3F241CFDD84343E0D52C05C10DA15C37E4B4040BB897A416EF8413B852512C1D45CC0001C949E7F77C5BFF022221C3E42D3C1763ACF5393A24FC0388B7B1E82A43BC00C41FA5472E2FE4148127B22928A5BC03699161014B445C0FF7ABCF1E7CAF1C10103000040010000000C000000949756B0091B41C0CFC1346AFD424BC0109D3C024F1CC3C1E0CFC79F4E3A16C06DF691401DF24FC028B253A07AF6E7C1684F7DE11B6C3B40C4BAFC73A75F3EC038C18EC0513ED641E690F20C0D0564409C3044D4E3F54DC0F015B39D4677CE414C995E715C40614098A5A801E82138C0D24D6C183D10F9415878CBC4B84D4E40CE6E49B89B49524090D463A3292ACE416D702D07D6D562C060FC8602B6294D40BBCFC1ACAE80FAC128FFEE2426A665C010FD32B40FFD4A40FB87B9E422EFFEC1E4F3648E40AD63C0C088D4C86D8945407845D843839201C280C1DAC2251B66C06232043C8DEC4240A8F9C035308CDF41201175131D8A5FC030D704BE4C0E2C40484745659286D2C1949756B0091B41C0CFC1346AFD424BC0109D3C024F1CC3C10103000060E610000001000000090000001E5727774DCA4EC04029D41071460BC0E9E4022744B5FDC1520BB2025D8254C05001F40F3C291AC05C0A3D15CCC300C2E842D184D07158408A4AA79F22FA50C0BBFB403130E9F4C1402C8C88DFC55B40A22C33203F0F4EC0A0716899E921EC4184D392299B416240E26E047EC53341405C30F5102D8EF84160B0F25DDB032940D0AB423939983A4000D4D97D43AE65C130F43B770C3A20C05409CC5E24B847401C9133F5982AF8C15E2FCC39761C65C028FDD3C5A3784A40806DF4A788CBD3C11E5727774DCA4EC04029D41071460BC0E9E4022744B5FDC10104000060E6100000070000000101000040307606CABF113BC0D8A459F4AA685440CC9035E09949E7C101010000400890DA6CAA8251C0F2757319FF5C40C09CB6DF548D22E641010100004020C87850634234C05B88D4D03F7C40C0F20F6BC6AA73F6C10101000040E06AFDAA52A955C0FA26146206B645C0AE0332C75B34F9410101000040489FE2BF44146440208101436B8738C0B9F08964C253F2C1010100004037DA2C10722555C050F38247F0BF1CC028FA3F365929D3C101010000407AC8E979BC556040EA6B59A444D554C034621B65FCB1F3410101000060E610000004D0FD0B41FC5A40680EAA3146EF5340143EDC40A129FC410106000060E61000000300000001030000400100000005000000083DA61F734F3440FB20D21D263244C0B830123D6C3AF841D897A0BFBBEB6040490BF5ECFDDE4AC074C098D49329EAC10CE9FA2BC4535940E847CF977ACF30C0B8B30D9E829AE6410B0C9993567264C0E86BE544BDAF274010BC7B1E3873C641083DA61F734F3440FB20D21D263244C0B830123D6C3AF8410103000040010000000D0000008D9844A5D70B54C0D069C457FD0829C0908AE16EA806C641E7A2D897FDE262C0E5C2A409F5AE42C0216AE45B8598FBC1F8BBE082B82A66C000C656215DAD53C000F05F5BCEF971413062CEBE1B2855C014EDC4B26FD342C0235DD39AA602FFC1E8271EAEE4BB40C0018128713D5856C0260877B2C42D01C29401E59DDA2756403BFA5C241EA153C044E7695E5FD6F8C1E46CF02DB4455E40F9CDA0220A6C48C0B0AE4E3ADFB3C7C1F02AB0DD27276340C84AAF76AAA55140007022B06A394EC1509CD3038E7439C04025158F9DD6FA3F347C0A53A6BDFBC1549AAC8E8E8D41C09E7C30531CC45340589924244CC0D6411B33A822529D5BC0D6B2B7162A145440AAE8FED038FC00C23B3A17B767F764C00AACD377F4CC50408CB4C4BB2198F6418D9844A5D70B54C0D069C457FD0829C0908AE16EA806C641010300004001000000090000001CF4F0CA9E4E5BC08302EC7CCE1148C074071D45487BE74179821125999C56C083928CA112BD54C0E6541CE728DFF6C1C87372DA5E106040E2886E6B21FF50C0085BA6E0CBF1F741C0E2C6FE9969174080ABFF02A26D00407E7D26622C36E6C1381111FC49C93D40381A450946FF4A400C7A3788B31BF041CC3F73F91FF841C03C4D3058D79D5440F04BAA2757BBD74114FAFC29925E43C040BA6E8544412740D0A0BADFD56BF4C102E2D6F7174A5CC020BE32F27AD61B40BA5C9E068403E8C11CF4F0CA9E4E5BC08302EC7CCE1148C074071D45487BE7410106000060E6100000090000000103000040010000000A000000473520ECB9A752C0B2B756FE629E4EC040D905148468B7C120FE0EB4D9773B40B841989DC10A55C020BB4C2F027EDF412EE21BBB8E615240BA1FA27F98E04DC00090ABCA6E7F72C158C0B3BDCE885D4049BDF40A99D84AC06B6505A0ADFFF7C15E32D0235C696040A8000886AF6E46C0028C688A4BCEFFC11AD1448BD94D664082CF4F7319A23FC0EC5AFAA59C5EF341FC800FCC3CD0544080AAD13B9D5D2E407CE754A2E176FA41E0E7914AA6B94B402420E5E7E34A4B4090E932F1E9C6FC41D8C5E38546FD3AC03601A22EE22D5340641C7A61F0FEF141473520ECB9A752C0B2B756FE629E4EC040D905148468B7C10103000040010000000B0000009D2F7E23FC4E51C0002CFB125FD14CC014122BB64AEEFB41101671B632873DC0E7316E8C98E44FC0AC112E9A2343ED4174DA6C6156E34640592D7A89EADB49C0F00F68FE8644D5414CE60D4B50C16040B86707559E244EC00C7345F21BFDEC4102EF9BCCF5A252402F1FECEF01DD44C08C0CF5C58095E2417E579EE5F81A6240DEEEB20C14C445C0BE078C2C9698E5C1C0231020565D5940C8A25B29DB9A25C069CD70ADFF82FFC1D84CD48A225157C0802BDF7E1026124098C0209528BBF741A4547FC72CA162C04068DF8BB2CD1240A4F666CF31F6EF414E91BFB7C1C264C0089F78B6954F3EC0D0779216CA1ECDC19D2F7E23FC4E51C0002CFB125FD14CC014122BB64AEEFB4101030000400100000005000000A43C194D16275F40AD9644470AB855C0A8EF67A8D815EA419E11BA0B6F7B6140537B1AD2222F50C0AE305180C6EDF941CAD915120E51644090D1F37A2D594B40BC4001FFBD63F6C1A8F4B57AE0DF5B4008168251AEF33540856F5571C0F9F8C1A43C194D16275F40AD9644470AB855C0A8EF67A8D815EA410103000040010000000B000000BBBB2CE5D58062C0A62A928C6CB455C00472BBE8BF1CFBC14879B8DFDA913D4074B9436B1CC550C0060E001880990042441896FE15C149407EDD8BCA5BB354C024A23225AC44E04120719E5D46A661406CB5BCC0B0193DC00052265F5D4EE4C10C1BCBDD2BD54340601B9822E3962440182AFC5A5368F941BAC5A3FFF7A15140A42B0F3C26484340483DFB3B812CD0C1947635C3852B514030C13F1E27A05340641719F689CBF0C1BCDC4E2A3F3B54C0FCC61473E21348401C7E1D7BC5E2F541DEA44821A95B60C070693717F48D294092E5660452800242684608EE3B6662C0B0313C7870792B400063489E5942B6C1BBBB2CE5D58062C0A62A928C6CB455C00472BBE8BF1CFBC10103000040010000000D00000013F7BD49EAE85CC030DD6D9B03BA1D40065FED8056C8F94190776C6CD6FF47C0DA281873AFF94BC0CC5E0789F393EA4180BC859B9FCE0B40C8531EE3A91A52C0345512A047DCF341A0E6D95B9B891B40294BAF9361D947C07E91D9B0426AF8C198ED56C0F0D64B402424781E92E936C0669B294D142B00421401868A7B3E534038510599CBBF37400AC32EA2F3C1E0C1DED942F7EE08514010B9A85000074240E051267EBD61FFC138A8D81454A361402AA07AEA3E525640ED8A831D924102C26C63677B2EE64540DCC3066E2E314E40CC6D551AAB3AE4410025B37B0A220DC0688B0AD8B87E4E40520228DD02C0F341BF000EE32CC663C0C0C20F46320340403E9301CE4C45E6C16B12FA8AD97E5AC080424672AA452440EC99D757A5FEE84113F7BD49EAE85CC030DD6D9B03BA1D40065FED8056C8F941010300004001000000040000007492EED46AAE5C40E647E2C640BA51C07C9DBF17CC610242841D8710178C554008A69AE945F63BC054D7736E7A90DFC1277D76E9919F65C036409115C24553405AB41446D69D02C27492EED46AAE5C40E647E2C640BA51C07C9DBF17CC610242010300004001000000050000008E933A015C4B5CC06E71D4C9097754C0624828BBDD94EFC1B05DD3753DBE65404291240A646040C040F8BABFD7FCAD41ACD81DF421DD63406C460D9DB3DE4F40029CE7DC69C3F841EE5F3E67A38660C09848FB3689AD4B40B8D002BE5FE4F0418E933A015C4B5CC06E71D4C9097754C0624828BBDD94EFC101030000400100000007000000705E19E3E34348C04007D455ECF2F3BF44816C46E588DBC1AAD63BCE275A5FC0C5E49CE7639350C0968387DAB332F2C19CE09F0E948E6140F06180D2AAE814C064A7D2157085EA4188E52A0AB28B4040E0F69541B54812403E8EE5E1CF2EECC144FDCBAD9A916140CC2091306F425140703FC105C738F04100322D560A8F1FC05419F55A06D832408043B61BAA4FB4C1705E19E3E34348C04007D455ECF2F3BF44816C46E588DBC101030000400100000009000000CACBF13F2BDC64C0702042FF6FBD3CC02426B3C60AB1FF41BCB1121ED9FA65407F2E883E1A874FC0C8FCD1A5CEAAE0C108DE89051D726540C30884C1883849C0E2543525820AF6C1583C639DBCE65B40E4251CA3C85D47402AAC972B2C800042B887EBC75DBE564066222046B23B5340105B8596988EEA411039C25A6D8B4840C83B5B6720244E4080BB8F62A2B1DE41F0848608D55F2340388B66D4026926405D67AA573D4C00C29AE7F0272D2660C0601B87048B89284080404D1E87B0FAC1CACBF13F2BDC64C0702042FF6FBD3CC02426B3C60AB1FF410102000060E6100000040000000E201A4603C862C0B508997DA00A4EC0BE7B370C5E4EF94130BAC472652F3740F029BDF0FA4943C0CA4AB61DC956F04158F3EF7F84BC65408D9D5443D73256C0F268025C651EF041E0140D06D2CB534080655C9FDF845140E0868FAD21DEB1410103000060E61000000100000005000000086F2E84F54035C0B6BBB5FB339649C038491DB6764D0042C04C2BB27FC24D405CDE7E9ADB6244409CFDD0901DEAEE4188A640232FF64EC06E1C125AF17453400891ED05B1C4F1415606F6DB6DD354C0C2956654729E424038D5D846D9D7D6C1086F2E84F54035C0B6BBB5FB339649C038491DB6764D00420102000060E61000000400000010861148148A64C08CAE7DD72C1637C08894B271CE40FE41B4A7FD8408C051C0FC876EFD202B3FC0D278F2881CA5FA410812471AC533394099BF137E60AA4CC0D825DAC64C0200420872DFDE07403740702ED1D36A631AC05EC0E5F60AE4F341':::GEOGRAPHY, '13:13:28.302372-08:10:00':::TIMETZ, NULL),(NULL, '01060000A0E61000000500000001030000800100000006000000AA58238866584AC050564A791AF828C0326C24947FDFF641844042863D7B43C06CD67475BD4F38C06CAE533A5FE1E94190C290C244FE3540F6F9B87B82F446C08029F267EC1D96C16CE1711F92CA5C40E079862D68312E40A0F163A837DEC2C16850A3F04C8F3AC06C28B4EAB7FC39401EFC0A0753A00142AA58238866584AC050564A791AF828C0326C24947FDFF6410103000080010000000B000000E51C23A7D58F50C018B0D81137942A4078D330A0CA91FA410C5C357B1C0745C09C50248C367938C0A06357B7F3CCFC4150DA44154BDC4640A080DEAA4F8824C010BDEC429415C0C128EFC70CBDE44A4070A6E161FB1929C0BE662418168B00C2F2F9E93DC6605240A42926DFA7DA33C0F413EB905ABCEFC1B84B77E6A9FF56400001DC0C44B520402410DB28474EEAC11E761E9546F654C04A51E31E7F7A5040CE6A894640D5F541EF770B8262295CC046B9EAEF986C53401AB669D6BD09E0C12EB4CCB6413E53C070E467AC73183C40024647605A7DF3419A67DBBB9C4C4EC0682903D7C3553840C2353421CF97EAC1E51C23A7D58F50C018B0D81137942A4078D330A0CA91FA410103000080010000000C000000015AF4CA547664C0AC259014DB5730C05C198098B980EC41D47BBAE865D643C021E0BB71077F48C04EB2A1624C7CF1410417C04934B54E401AF50ECBAA5956C03C8767584BD6DAC1E41C9EA1A6EB5C4002AEF77BFB0356C0807E86B9A69DBC4100CB89BB45E32EC0A002415E077A1C404C2DB7684DB2F641609B28A2194B1AC0B0B89E8CD086244056F581276163E4C10060AB6CCBC9AD3F5C5CB56FF2303F4068E4641E247BEDC130BA13E0E85F2EC00C8B450935BF5240B8E84599FF2ED241026EB8991CFA4EC090F612242B815440AAABCA25A0A3F4416EEF909FD12B4DC0C4E78EBB28584E4094F18A92691501C25875E87B0ED662C00CEB1ADE59274040A2BD687D698C0042015AF4CA547664C0AC259014DB5730C05C198098B980EC4101030000800100000008000000244DA5084B2D56C03796C46B193352C0242A16CB5009ED413635359B043D4FC0E8F62566392146C0506574C37D8BEF41D8D77401C6E532C0525E7BE64C4947C0068BED7901B3E4C19E3A3D150E245040E8A107602A722E4080CABA60DAC2EFC1849A10DE73074340FA7E4EEFC4F454405DA8E8604F0BF5C192CF9333CBFA5DC008EEADA6C8564B400699B9DFB57EE5C1FD03BA290E3166C0C822BD2427A732402AEC2B59FFCEE0C1244DA5084B2D56C03796C46B193352C0242A16CB5009ED410103000080010000000D00000063216EDB5CAF65C06423D7A023364BC028254B9B6C47E2C1B027569557C23C4060071E1353792740702CD924D676D64140C691F290784C4060CC1943286308402CC6391F7680EB4126C69D0991536140000FE8B0E3BA07C06487F271E541E841D63B8213A01F6040F06F5B0173D23F409198B23C6763FFC1FC52865A96326540D48891CD57C148408008CF49878CFB41864C5767AED26040A8CD6A3D5C7A49408D6D6A81E61DFAC1106AF08CAA0B5940A64D992A557754406C0A6B07E76BF3C160DC8EBADBE134C0FCB889EA8DED4C402031EE35B2E3FCC1BD9E5E23958553C04839535B8B3748407463EC321AD4E1C1A2650B44442E66C01C9569E2D1E641405C37FF7F31DEF3C100F973535D30E93FE488414B52C93C401E065242B1F8F34163216EDB5CAF65C06423D7A023364BC028254B9B6C47E2C1':::GEOGRAPHY, '01:04:53.854855-07:05:00':::TIMETZ, '\x75cd0a622b440bd3':::BYTES)], '1971-11-27':::DATE)
				AS col_27124
		FROM
			with_5405 AS cte_ref_1531
		ORDER BY
			cte_ref_1531.col_27118, cte_ref_1531.col_27118, cte_ref_1531.col_27118
		LIMIT
			66:::INT8;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-ddl.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-14881

```
05:28:51 test_impl.go:312: test failure: 	sqlsmith.go:265,sqlsmith.go:304,test_runner.go:875: error: pq: internal error: unexpected error from the vectorized engine: error decoding 369 bytes: couldn't decode type anyelement
```
Not a release blocker.

--- a/pkg/sql/logictest/testdata/logic_test/tuple
+++ b/pkg/sql/logictest/testdata/logic_test/tuple
@@ -1217,3 +1217,25 @@ SELECT (CASE WHEN false THEN (ROW(1) AS a) ELSE (ROW(2) AS b) END).a;
 
 statement error could not identify column \"b\" in tuple{int AS a}
 SELECT (CASE WHEN false THEN (ROW(1) AS a) ELSE (ROW(2) AS b) END).b;
+
+subtest end
+
+# Regression test for #79484. Verify that type-checking is smart enough
+# to handle tuples of arrays of tuples in the VALUES clause.
+subtest 79484
+
+query T
+SELECT * FROM (
+  VALUES
+    ((ARRAY[15181:::INT8], ARRAY[]:::RECORD[])),
+    (
+      (ARRAY[1534:::INT8], ARRAY[('infinity':::DATE, 'cat':::TEXT)])
+    ),
+    (NULL)
+) AS tab(col)
+----
+({15181},{})
+({1534},"{""(infinity,cat)""}")
+NULL
+
+subtest end

--- a/pkg/sql/opt/optbuilder/values.go
+++ b/pkg/sql/opt/optbuilder/values.go
@@ -78,7 +78,7 @@ func (b *Builder) buildValuesClause(
 			if typ := texpr.ResolvedType(); typ.Family() != types.UnknownFamily {
 				if colTypes[colIdx].Family() == types.UnknownFamily {
 					colTypes[colIdx] = typ
-				} else if rightHasMoreSpecificTuple(colTypes[colIdx], typ) {
+				} else if moreSpecific, _ := rightHasMoreSpecificTuple(colTypes[colIdx], typ); moreSpecific {
 					// This condition handles the case when an earlier expression in the
 					// VALUES clause is an array of AnyTuple, but a later expression is an
 					// array of a more specific tuple type.
@@ -140,37 +140,35 @@ func reportValuesLenError(expected, actual int) {
 		expected, actual))
 }
 
-// rightHasMoreSpecificTuple returns true if the left and right types are
-// equivalent, but the right type is constrained by a more specific nested
-// tuple than the left type.
-func rightHasMoreSpecificTuple(left, right *types.T) bool {
+// rightHasMoreSpecificTuple returns two values. The first return parameter is
+// true if the left and right types are equivalent, but the right type is
+// constrained by a more specific nested tuple than the left type. The second
+// return parameter is true if the types are equivalent.
+func rightHasMoreSpecificTuple(left, right *types.T) (isMoreSpecific bool, isEquivalent bool) {
 	if left.Family() != right.Family() {
-		return false
+		return false, false
 	}
 	if left.Family() == types.ArrayFamily && right.Family() == types.ArrayFamily {
 		return rightHasMoreSpecificTuple(left.ArrayContents(), right.ArrayContents())
 	}
 	if left.Family() == types.TupleFamily && right.Family() == types.TupleFamily {
 		if right == types.AnyTuple {
-			return false
+			return false, true
 		} else if left == types.AnyTuple {
-			return true
+			return true, true
 		} else if len(left.TupleContents()) != len(right.TupleContents()) {
-			return false
+			return false, false
 		}
-		ret := true
+		allEquivalent := true
+		atLeastOneMoreSpecific := false
 		for i, leftElem := range left.TupleContents() {
 			rightElem := right.TupleContents()[i]
-			// Only recurse if we are dealing with a container type.
-			if leftElem.Family() == types.ArrayFamily || leftElem.Family() == types.TupleFamily {
-				ret = ret && rightHasMoreSpecificTuple(leftElem, rightElem)
-			} else {
-				ret = ret && leftElem.Equivalent(rightElem)
-			}
+			elemIsMoreSpecific, elemIsEquivalent := rightHasMoreSpecificTuple(leftElem, rightElem)
+			allEquivalent = allEquivalent && elemIsEquivalent
+			atLeastOneMoreSpecific = atLeastOneMoreSpecific || elemIsMoreSpecific
 		}
-		return ret
+		return allEquivalent && atLeastOneMoreSpecific, allEquivalent
 	}
-	// For non-tuples and non-arrays, we don't want the right type to get
-	// preference over the left type.
-	return false
+	// At this point, both left and right are neither tuples nor arrays.
+	return false, left.Equivalent(right)
 }

